Security chapter

ScottHelme opened the pull request:
Creating a draft PR for feedback (@rviscomi) and collab (@arturjanc).

ScottHelme commented:
@rviscomi is acceptable to include the data as tables like I have been doing or should I be lifting out all of the metrics and writing them up?

rviscomi commented:
@ScottHelme thanks for writing this up! Yes, tables are fine. We also have a data viz theme for Sheets that we can apply to any charts you may want to embed. If so, just leave a placeholder in the doc like `<insert bar chart of 08_03>`

rviscomi commented:
@ScottHelme @arturjanc just checking in. Do you think you could finish the draft by the end of the week? 🙏

cc reviewers: @paulcalvano @bazzadp @ghedo @ndrnmnn

ghedo commented:
Looks good so far, but haven't reviewed the whole thing yet.

Regarding the tables, it seems to me it would be better to split mobile and desktop into separate tables to make it easier to look at the data  In some cases (e.g. RSA vs. ECDSA) it would make the table simpler as well I think.

ghedo commented:
Also, the raw counts in the tables are probably not that useful either, and just having percentages seems like it would be easier to look at that (as well as allow making pie charts and the like).

bazzadp commented:
> Regarding the tables, it seems to me it would be better to split mobile and desktop into separate tables to make it easier to look at the data In some cases (e.g. RSA vs. ECDSA) it would make the table simpler as well I think.

> Also, the raw counts in the tables are probably not that useful either, and just having percentages seems like it would be easier to look at that (as well as allow making pie charts and the like).

Agreed. Our just have the mobile and desktop as columns, rather than two tables. Check out the HTTP/2 chapter for how I handled this: https://github.com/HTTPArchive/almanac.httparchive.org/blob/master/src/content/en/2019/http2.md

ghedo commented on src/content/en/2019/security.md:
```diff
@@ -0,0 +1,460 @@
+# Chapter 8: Security
+
+## Intro
+The Security chapter of the Web Almanac looks at the progress made in securing the web. With security and privacy becoming increasingly more important online there has been and increase in the availability of features to protect site operators and users. We're going to look at the adoption of these new features across the Web.
```
```suggestion
The Security chapter of the Web Almanac looks at the progress made in securing the web. With security and privacy becoming increasingly more important online there has been an increase in the availability of features to protect site operators and users. We're going to look at the adoption of these new features across the Web.
```

ghedo commented on src/content/en/2019/security.md:
```diff
@@ -0,0 +1,460 @@
+The rise of Let's Encrypt has been meteoric after their launch in early 2016, since then they've become one of the top certificate issuers in the world. The availability of free certificates and automated tooling has been critically important to the adoption of HTTPS on the Web and Let's Encrypt certainly had a significant part to play in both of those.
+
+### Authentication key type
+Alongside the important requirement to use HTTPS is the requirement to also use a good configuration. With so many configuration options and choices to make, this is a careful balance. Looking first of all the keys used for authentication purposes, we still see a large % of the Web using RSA.
```
```suggestion
Alongside the important requirement to use HTTPS is the requirement to also use a good configuration. With so many configuration options and choices to make, this is a careful balance. Looking first of all at the keys used for authentication purposes, we still see a large % of the Web using RSA.
```

That last sentence reads a bit weird...

ghedo commented on src/content/en/2019/security.md:
```diff
@@ -0,0 +1,460 @@
+
+<br/>
+
+Whilst ECDSA keys are stronger and demonstrate better performance than their RSA counterparts, concerns around backwards compatibility do prevent some website operators from migrating. 
```
```suggestion
Whilst ECDSA keys are stronger, which allows the use of smaller keys, and demonstrate better performance than their RSA counterparts, concerns around backwards compatibility do prevent some website operators from migrating. 
```

Also worth noting that support around ECDSA is not that great, e.g. many LE clients will only do RSA by default

ghedo commented on src/content/en/2019/security.md:
```diff
@@ -0,0 +1,460 @@
+
+<br/>
+
+Looking at the % of TLS requests that utilised Forward Secrecy, we can see that support is tremendous. We'd expect that the continuing increase in the adoption of TLSv1.3 will further increase these numbers.
```
```suggestion
Looking at the % of TLS requests that provide Forward Secrecy, we can see that support is tremendous. We'd expect that the continuing increase in the adoption of TLSv1.3 will further increase these numbers.
```

ghedo commented on src/content/en/2019/security.md:
```diff
@@ -0,0 +1,460 @@
+Whilst ECDSA keys are stronger and demonstrate better performance than their RSA counterparts, concerns around backwards compatibility do prevent some website operators from migrating. 
+
+### Forward Secrecy
+The importance of Forward Secrecy is now well understood. It was introduced as an optional configuration in 2008 with TLSv1.2 and has become mandatory in 2018 with TLSv1.3 requiring it.
```
Worth explaining briefly what forward secrecy is? e.g.

```suggestion
The importance of Forward Secrecy, a property of some key exchange algorithms that prevents each connection to a server from being exposed in case of a compromise of the server certificate's private key, is now well understood. It was introduced as an optional configuration in 2008 with TLSv1.2 and has become mandatory in 2018 with TLSv1.3 requiring it.
```

ScottHelme commented:
I've incorporated all of those changes @ghedo, I made a mistake on the commit and lost them so had to add them back in! How is the chapter looking in terms of length? This is just a first cut to get the basic information and layout in there, but I can go into more detail than this. Would you say it's needs expanding more?

ScottHelme commented:
@bazzadp had a crack at including your comments, I think I caught most of them. Also updated tables and added a graph placeholder. 

Will probably give it one more pass this week for final visual tweaks like graphs etc... but I think it's close now.

bazzadp commented:
Cool. Will have another look.

You couldn't remove that last commit to the 2019.json file would you? I've done a bulk update in https://github.com/HTTPArchive/almanac.httparchive.org/pull/272 including your Twitter handle and want to avoid any Git merge messiness.

bazzadp commented:
I still think the tables could be formatted better for mobile and desktop. Very difficult to read as it is at the moment.

For example this:

| Client  | HTTP vs HTTPS | Sites using this scheme (%) |
|---------|---------------|-----------------------------|
| desktop | HTTPS         | 60.25%                      |
| mobile  | HTTPS         | 54.33%                      |
| desktop | HTTP          | 39.75%                      |
| mobile  | HTTP          | 45.67%                      |

Could be displayed as:

| Protocol     | Desktop | Mobile |
|---------|---------------|-----------------------------|
| HTTPS  | 60.25%           |    54.33%                     |
| HTTP    | 39.75%          |   45.67%                 |

Which allows easier comparison across Protocols and Client Types. As well as making the table a good bit smaller.

This should then be repeated similarly for all the other tables in a similar format (Desktop and Mobile as columns in a consistent order).

Other than that there's still a lot of unresolved comments above. I still think they have merit but feel free to close them if you disagree.

ScottHelme commented:
Removed the contributor commit. 

Will keep reading through comments and adding what I can, it's more a time constraint than an agreeing with the comments constraint :)

Will also try and change all of the tables, it does look better.

rviscomi commented:
@ScottHelme this is in great shape, we're almost there. @arturjanc and I spoke over email and he hopes to spend some time on this chapter over the weekend. Artur, highest priority would be to work on a Conclusion section to wrap it all up and fill in any missing sections. We'll be launching this next week! Thanks both!

arturjanc commented:
As a small update, I'm working on the conclusion section now and should have a PR by EOD.

arturjanc commented:
@ScottHelme did all the real work here, but I wrote some words in https://github.com/HTTPArchive/almanac.httparchive.org/commit/f7e4539f9d6aeedf9ed07aeda878feb19b4ac3d0 to provide a high-level overview of the data summarized in the chapter - PTAL.

Looking at our original plan from https://github.com/HTTPArchive/almanac.httparchive.org/issues/10#issuecomment-501876182 the current content seems to cover all the major features we wanted to measure (there are a few smaller ones such as CORP and COOP, but they have barely any adoption so I don't think they'd add much at this point.)

rviscomi commented:
Looks great @arturjanc! Wonderful conclusion.

I'll look into the merge conflict so we can merge this.

rviscomi commented:
@ScottHelme this is in great shape. @bazzadp left a bunch of peer review feedback that need to be addressed before closing this out. We're launching in a matter of days so the soonest you can resolve these comments would be a huge help. 🙏

rviscomi commented:
🛎 friendly ping @ScottHelme @arturjanc. We're very very close to launch so it'd be great to get this resolved ASAP! Thank you for your understanding!

rviscomi commented:
🛎 🛎 another ping. Sorry to nag. @arturjanc @ScottHelme do you have time to close this out?

bazzadp commented:
@arturjanc @ScottHelme since the major feedback is my own, if you’re happy we can merge this as is and I can take this as the first part of copy editing and include you as approval on that copyediting PR.

@rviscomi then you or someone else can do a second round of copyediting as someone unrelated to the chapter but hopefully that’ll be light since I can do visualisation and the other major parts of copy editing.

Can you all let me know if you’re not OK with that by the end of the day and I’ll do it tomorrow.

arturjanc commented:
I think @ScottHelme's text (and hopefully the conclusion) are mergeable as is. @bazzadp, if you're willing to integrate your suggested changes in a follow-up PR I can review it today. From looking at them they all seem reasonable, so unless Scott objects I'd expect this is fine.

bazzadp commented:
It won't be today (day job and all that) but hopefully late this evening/early tomorrow am.

rviscomi commented:
@bazzadp great idea, let's keep this moving.
